File: generator/src/java/darkness/generator/api/effects/RGBFade.java
package darkness.generator.api.effects;

import darkness.generator.api.BulbSet;

import java.awt.Color;

public class RGBFade extends EffectBase {
    private final Color startColor;
    private final Color endColor;
    private final BulbSet bulbSet;
    private final int frames;
    private boolean cancelled = false;
    private final float delta_r, delta_g, delta_b;

    public RGBFade(BulbSet bulbSet, Color endColor, int frames) {
        this.startColor = bulbSet.getColor();
        this.endColor = endColor;
        this.bulbSet = bulbSet;
        this.frames = frames;
        this.delta_r = (float)(endColor.getRed() - startColor.getRed()) / (float)frames;
        this.delta_g = (float)(endColor.getGreen() - startColor.getGreen()) / (float)frames;
        this.delta_b = (float)(endColor.getBlue() - startColor.getBlue()) / (float)frames;
    }

    @Override
    public void run() {
        for(int frame = 1; frame <= frames; frame++) {
            if(cancelled) {
                return;
            }
            int red  = startColor.getRed() + (int)(delta_r*frame);
            int green  = startColor.getGreen() + (int)(delta_g*frame);
            int blue = startColor.getBlue() + (int)(delta_b*frame);

            set(bulbSet, red, green, blue);
            next();
        }
    }

    @Override
    public void cancel() {
        cancelled = true;
    }

    @Override
    public String toString() {
        return "Effect RGBFade on " + bulbSet + " from color: " + startColor + " to color: " + endColor + " over " + frames + " frames.";
    }
}

File: generator/src/kotlin/darkness/generator/api/effects/RGBFade.kt
package darkness.generator.api.effects

import darkness.generator.api.BulbSet

import java.awt.Color

class RGBFade(
    private val bulbSet: BulbSet,
    private val endColor: Color,
    private val frames: Int
) : EffectBase() {
    private val startColor = bulbSet.color
    private val delta_r: Float
    private val delta_g: Float
    private val delta_b: Float

    init {
        this.delta_r = (endColor.red - startColor.red).toFloat() / frames.toFloat()
        this.delta_g = (endColor.green - startColor.green).toFloat() / frames.toFloat()
        this.delta_b = (endColor.blue - startColor.blue).toFloat() / frames.toFloat()
    }

    override fun run() {
        var frame = 1
        while (frame <= frames && !isCancelled) {
            val red = startColor.red + (delta_r * frame).toInt()
            val green = startColor.green + (delta_g * frame).toInt()
            val blue = startColor.blue + (delta_b * frame).toInt()

            set(bulbSet, red, green, blue)
            next()
            frame++
        }
    }

    override fun toString() = "Effect RGBFade on $bulbSet from color: $startColor to color: $endColor over $frames frames."
}
